src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs

```diff
@@ -211,5 +210,153 @@ private IEnumerable<AvailableCodeAction> ConvertToAvailableCodeAction(IEnumerabl
                 return new[] { new AvailableCodeAction(action) };
             });
         }
+
+        // Mapping means: each mapped item has one document that has one code fix provider and it's corresponding diagnostics.
+        // If same document has multiple codefixers (diagnostics with different fixers) will them be mapped as separate items.
+        protected async Task<ImmutableArray<DocumentWithFixProvidersAndMatchingDiagnostics>> GetDiagnosticsMappedWithFixAllProviders(FixAllScope scope, string fileName)
+        {
+            ImmutableArray<DocumentDiagnostics> allDiagnostics = await GetCorrectDiagnosticsInScope(scope, fileName);
+
+            var mappedProvidersWithDiagnostics = allDiagnostics
+                .SelectMany(diagnosticsInDocument =>
+                    DocumentWithFixProvidersAndMatchingDiagnostics.CreateWithMatchingProviders(codeFixesForProject.GetAllCodeFixesForProject(diagnosticsInDocument.ProjectId), diagnosticsInDocument));
+
+            return mappedProvidersWithDiagnostics.ToImmutableArray();
+        }
+
+        private async Task<ImmutableArray<DocumentDiagnostics>> GetCorrectDiagnosticsInScope(FixAllScope scope, string fileName)
+        {
+            switch (scope)
+            {
+                case FixAllScope.Solution:
+                    var documentsInSolution = Workspace.CurrentSolution.Projects.SelectMany(x => x.Documents).Select(x => x.FilePath).ToImmutableArray();
+                    return await diagnostics.GetDiagnostics(documentsInSolution);
+                case FixAllScope.Project:
+                    var documentsInProject = Workspace.GetDocument(fileName).Project.Documents.Select(x => x.FilePath).ToImmutableArray();
+                    return await diagnostics.GetDiagnostics(documentsInProject);
+                case FixAllScope.Document:
+                    return await diagnostics.GetDiagnostics(ImmutableArray.Create(fileName));
+                default:
+                    throw new NotImplementedException();
+            }
+        }
+
+        protected async Task<(Solution Solution, IEnumerable<FileOperationResponse> FileChanges)> GetFileChangesAsync(Solution newSolution, Solution oldSolution, string directory, bool wantTextChanges, bool wantsAllCodeActionOperations)
```

Comment:
Refactored this down from CodeActionService so that FixAllService can use it too.